flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/GSFileSystemFactory.java

```diff
@@ -0,0 +1,136 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.fs.gs;
+
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.FileSystemFactory;
+import org.apache.flink.runtime.util.HadoopConfigLoader;
+import org.apache.flink.util.Preconditions;
+
+import com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+
+/**
+ * Implementation of the Flink {@link org.apache.flink.core.fs.FileSystemFactory} interface for
+ * Google Storage.
+ */
+public class GSFileSystemFactory implements FileSystemFactory {
+
+    private static final String SCHEME = "gs";
+
+    private static final String HADOOP_CONFIG_PREFIX = "fs.gs.";
+
+    private static final String[] FLINK_CONFIG_PREFIXES = {"gs.", HADOOP_CONFIG_PREFIX};
+
+    private static final String[][] MIRRORED_CONFIG_KEYS = {};
+
+    private static final String FLINK_SHADING_PREFIX = "";
+
+    public static final ConfigOption<String> WRITER_TEMPORARY_BUCKET_NAME =
+            ConfigOptions.key("gs.writer.temporary.bucket.name")
+                    .stringType()
+                    .defaultValue(GSFileSystemOptions.DEFAULT_WRITER_TEMPORARY_BUCKET_NAME)
+                    .withDescription(
+                            "This option sets the bucket name used by the recoverable writer to store temporary files. "
+                                    + "If empty, temporary files are stored in the same bucket as the final file being written.");
+
+    public static final ConfigOption<String> WRITER_TEMPORARY_OBJECT_PREFIX =
+            ConfigOptions.key("gs.writer.temporary.object.prefix")
+                    .stringType()
+                    .defaultValue(GSFileSystemOptions.DEFAULT_WRITER_TEMPORARY_OBJECT_PREFIX)
+                    .withDescription(
+                            "This option sets the prefix used by the recoverable writer when writing temporary files. This prefix is applied to the "
+                                    + "final object name to form the base name for temporary files.");
```

Comment:
The main reason to consider adding this knob is to allow one to use google storage's [object lifecycle management](https://cloud.google.com/storage/docs/lifecycle) to assign TTLs to temporary files. In google storage, TTLs can only be assigned at the bucket level, not the individual object level. So it would be impossible to assign TTLs to temporary objects that reside in the same bucket as the "final" objects without also applying the TTL to the final objects.

I do think this could be useful, especially when I make the changes you've suggested to make the deletion of temporary files more focused; the TTL mechanism might be the only way orphaned temporary files would ever get deleted. What do you think?

Reply:
Thanks for the explanation. I agree with you that being able to specify different TTLs for temporal/final objects can be useful, which makes a good reason for making the temporary bucket configurable.

What about the temporal object prefix? Is it necessary to make that configurable as well?

Reply:
It's not really necessary, I suppose. The use case I was envisioning here was this: When you specify a dedicatedt bucket to hold in-progress temporary blobs, the ".inprogress" prefix is sort of redundant. So this lets you set it to an empty string and have the files just appear directly in the root of the bucket, i.e. the files would look like:

```/foo/bar/7b342499-6918-48f0-bcf9-11cf2bc18c51```

... instead of ...

```/.inprogress/foo/bar/7b342499-6918-48f0-bcf9-11cf2bc18c51```

Admittedly, this isn't really that important to be able to do. So I'm fine with removing this option if that's what you'd prefer.

Also, this reminds me of a change I'm planning to submit with the next batch, which is to have the generated temporary blob names include both the final bucket name and the final object name. Currently, the temporary blob names only include the final object name, but not the bucket name. So what is now:

```/.inprogress/foo/bar/7b342499-6918-48f0-bcf9-11cf2bc18c51```

... would become:

```/.inprogress/bucket_name/foo/bar/7b342499-6918-48f0-bcf9-11cf2bc18c51```

This keeps temporary blobs properly separated in the event that a) a dedicated bucket for temporary blobs is used and b) a blob with the same name (i.e. /foo/bar) is written into two different buckets by a StreamingFileSink at the same time.

(And, it would actually technically still work *without* adding the bucket name, because the UUIDs would be distinct, but it just seemed to be potentially confusing to commingle the temporary blobs for different writes in the same storage tree.

Reply:
I'd be in favor of not introducing this option if it's not absolutely necessary. The more options we provide to the users, the more constraints we put on future developing and maintaining.

As for including the bucket names in temporary blob names, I think that's a good idea. Since the temporary blobs are not meant to be manipulated by users directly, it would be nice to carry more information with the blob names.

Moreover, I wonder whether we should also include the index of temporary blobs, in addition to the UUID. That might be useful for cleaning up the resumables. I'll explain in the other comment.

Reply:
Config for temporary-object prefix removed in [2dc286c](https://github.com/apache/flink/pull/15599/commits/2dc286c697fff7ad6adb68c2a0e6de40d1c20bfe).